### ISSUE TITLE
Decodig sampled SLATs vs Encoded SLATs

### ISSUE THREAD

**Issue by ariel415el** [microsoft/TRELLIS.2#53]
Hello and thanks for your great twork!

I'm trying to test the reconstruction quality of the shape-VAEs (I'm reconstructing a textureless mesh).
When running **example.py** and saving the intermediate mesh that is decoded by the shape decoder in
https://github.com/microsoft/TRELLIS.2/blob/903bfcf51af09b3fc6c408d2c7f2335febab7f81/trellis2/pipelines/trellis2_image_to_3d.py#L470
it looks resonably well (see below - image2mesh-decoder output ).
Unfortunately, when I'm using the shape-encoder (even on the same mesh created by image2mesh - image2mesh-full below) and decoding the SLATs (with the code below - no postprocess), i'm getting very voxelized mesh (encode-decode  below).
what is the difference between the two? How come sampled SLATS are so different from encoded ones wasn't the diffusion model trained to sample encoded SLATs?

Images and code are below
Thanks for helping

| input-image        | image2mesh-decoder output  | image2mesh-full   | encode-decode   |
|-----------:|:----------:|:-------------:|:-------------:|
|<img width="403" height="450" alt="Image" src="https://github.com/user-attachments/assets/19679c36-3065-4b6b-ad5b-b66eb0e3c1aa" />| <img width="403" height="740" alt="Image" src="https://github.com/user-attachments/assets/6245bbe5-a4d4-41f4-9a5f-49c9d6060f94" />  | <img width="403" height="740" alt="Image" src="https://github.com/user-attachments/assets/00cd0f52-8dde-4976-bdef-8cd0ecf6fd88" />  | <img width="403" height="740" alt="Image" src="https://github.com/user-attachments/assets/460a9abd-fa86-40a6-85b9-be9fa3bbf224" />

```
import torch
import trimesh
import o_voxel
from trellis2 import models
from trellis2.modules.sparse import SparseTensor
from trellis2.representations import MeshWithVoxel


def load_shape_models():
    shape_enc = models.from_pretrained("microsoft/TRELLIS.2-4B/ckpts/shape_enc_next_dc_f16c32_fp16")
    shape_dec = models.from_pretrained("microsoft/TRELLIS.2-4B/ckpts/shape_dec_next_dc_f16c32_fp16")
    for model in [shape_enc, shape_dec]:
        model.cuda().eval()
    
    return shape_enc, shape_dec


def normalize_mesh(mesh):
    """Normalize mesh vertices to unit cube."""
    vertices = torch.from_numpy(mesh.vertices).float()
    faces = torch.from_numpy(mesh.faces).long()
    
    center = (vertices.min(0)[0] + vertices.max(0)[0]) / 2
    scale = 0.99999 / (vertices.max(0)[0] - vertices.min(0)[0]).max()
    vertices = (vertices - center) * scale
    
    return vertices, faces


def encode_decode_shape(encoder, decoder, vertices, faces, resolution=512):
    voxel_indices, dual_vertices, intersected = o_voxel.convert.mesh_to_flexible_dual_grid(
        vertices, faces, grid_size=resolution,
        aabb=[[-0.5, -0.5, -0.5], [0.5, 0.5, 0.5]],
    )
    
    coords = torch.cat([torch.zeros(len(voxel_indices), 1, dtype=torch.int32), voxel_indices], dim=1)
    vertices_sparse = SparseTensor(dual_vertices.cuda(), coords.cuda())
    intersected_sparse = SparseTensor(intersected.float().cuda(), coords.cuda())
    
    with torch.no_grad():
        latent = encoder(vertices_sparse, intersected_sparse)
        decoder.set_resolution(resolution)
        meshes, subs = decoder(latent, return_subs=True)
    
    return meshes[0], subs


def postprocess_with_default_texture(mesh, subs, resolution):
    pbr_attr_layout = {
        'base_color': slice(0, 3),
        'metallic': slice(3, 4),
        'roughness': slice(4, 5),
        'alpha': slice(5, 6),
    }
    
    mesh.fill_holes()
    
    # Create default white material
    sub = subs[-1]
    num_voxels = sub.coords.shape[0]
    default_attrs = torch.ones(num_voxels, 6, device=sub.coords.device, dtype=torch.float32)
    default_attrs[:, 3] = 0.0  # metallic
    default_attrs[:, 4] = 0.5  # roughness
    default_attrs[:, 5] = 1.0  # alpha
    
    mesh_with_voxel = MeshWithVoxel(
        mesh.vertices, mesh.faces,
        origin=[-0.5, -0.5, -0.5],
        voxel_size=1 / resolution,
        coords=sub.coords[:, 1:],
        attrs=default_attrs,
        voxel_shape=torch.Size([*sub.shape, *sub.spatial_shape]),
        layout=pbr_attr_layout
    )
    
    mesh_with_voxel.simplify(16777216)
    
    glb = o_voxel.postprocess.to_glb(
        vertices=mesh_with_voxel.vertices,
        faces=mesh_with_voxel.faces,
        attr_volume=mesh_with_voxel.attrs,
        coords=mesh_with_voxel.coords,
        attr_layout=mesh_with_voxel.layout,
        voxel_size=mesh_with_voxel.voxel_size,
        aabb=[[-0.5, -0.5, -0.5], [0.5, 0.5, 0.5]],
        decimation_target=1000000,
        texture_size=4096,
        remesh=True,
        remesh_band=1,
        remesh_project=0,
        verbose=True
    )
    return glb


def save_mesh(mesh, output_path):
    trimesh.Trimesh(
        vertices=mesh.vertices.cpu().numpy(),
        faces=mesh.faces.cpu().numpy(),
    ).export(output_path)


def main():
    shape_enc, shape_dec = load_shape_models()
    
    mesh_path = "/common_data/workspace/ariel/repos/Hunyuan3D-2.1/assets/cpt_america_out.glb"
    mesh = trimesh.load(mesh_path)
    if isinstance(mesh, trimesh.Scene):
        mesh = list(mesh.geometry.values())[0]
    vertices, faces = normalize_mesh(mesh)
    
    resolution = 1024
    shape_mesh, subs = encode_decode_shape(shape_enc, shape_dec, vertices, faces, resolution=resolution)
    save_mesh(shape_mesh, "reconstructed_raw_shape.glb")

    glb_mesh = postprocess_with_default_texture(shape_mesh, subs, resolution)
    glb_mesh.export("reconstructed_mesh_postprocessed.glb")
    


if __name__ == "__main__":
    main()
```

**Comment by animemory** [comment]
same here, have u got any idea?

**Comment by yuyujunjun** [comment]
After o_voxel.convert.mesh_to_flexible_dual_grid, transform vertices position to the space relative grid
dual_vertices = dual_vertices * res - voxel_indices

**Comment by ariel415el** [comment]
Thanks @yuyujunjun it seems to help.
However, the output mesh does not seem to have the same holes as in the image2mesh-decoder output I uploaded before.
Do you have an idea where does these holes come from

**Comment by yuyujunjun** [comment]
I don't know where does these holes comes from. I assume there may be some precision issues. The original code contains fill_holes method which may be employed to address this.

**Comment by ydayda** [comment]
I test the code with fix @yuyujunjun :  dual_vertices = dual_vertices * res - voxel_indices  
but I found the to_glb  function  seems has loss! Just the same the [example.py ](https://github.com/microsoft/TRELLIS.2/blob/903bfcf51af09b3fc6c408d2c7f2335febab7f81/o-voxel/examples/ovox2glb.py#L38)
, which I mentioned at [this comment](https://github.com/microsoft/TRELLIS.2/issues/124). 
**the generation 3D model will be much better if fix this  issue!**  @ariel415el @animemory 

EDIT: seems caused by [this](https://github.com/microsoft/TRELLIS.2/blob/903bfcf51af09b3fc6c408d2c7f2335febab7f81/o-voxel/o_voxel/postprocess.py#L185), too slow! if skip to the simplify!
<img width="2206" height="1057" alt="Image" src="https://github.com/user-attachments/assets/6db37085-35b7-4af0-9322-ff338b9fc221" />

**Comment by animemory** [comment]
> I test the code with fix [@yuyujunjun](https://github.com/yuyujunjun) : dual_vertices = dual_vertices * res - voxel_indices but I found the to_glb function seems has loss! Just the same the [example.py ](https://github.com/microsoft/TRELLIS.2/blob/903bfcf51af09b3fc6c408d2c7f2335febab7f81/o-voxel/examples/ovox2glb.py#L38) , which I mentioned at [this comment](https://github.com/microsoft/TRELLIS.2/issues/124). **the generation 3D model will be much better if fix this issue!** [@ariel415el](https://github.com/ariel415el) [@animemory](https://github.com/animemory)
> 
> EDIT: seems caused by [this](https://github.com/microsoft/TRELLIS.2/blob/903bfcf51af09b3fc6c408d2c7f2335febab7f81/o-voxel/o_voxel/postprocess.py#L185), too slow! if skip to the simplify! <img alt="Image" width="2000" height="1057" src="https://private-user-images.githubusercontent.com/250907484/555120615-6db37085-35b7-4af0-9322-ff338b9fc221.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzIxNjQwNjcsIm5iZiI6MTc3MjE2Mzc2NywicGF0aCI6Ii8yNTA5MDc0ODQvNTU1MTIwNjE1LTZkYjM3MDg1LTM1YjctNGFmMC05MzIyLWZmMzM4YjlmYzIyMS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwMjI3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDIyN1QwMzQyNDdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yODcyZjkwNjI3NWQ4YjVkYzQ1YTMyOTcxMTMyYTVhM2ZhM2MzZjUyNWMwOTViMDU5OGI4NDE5OGQ4NWNmNGQxJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.6awtQqrxgM5ge_vEsOQRu40P3xYLRZrQ8zq_6vw9RuA">

thanks for your detailed test! is there any alternative for such simplifying method? Why there huggingface tests seems not that slow?
